File: code/ex24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        #Deal with length <= 1
        if head == None:
            return []
        elif head.next == None:
            return head

        #Deal with the first two nodes
        p = head
        q = p.next
        p.next = q.next
        q.next = p
        head = q
        
        #If there are only two nedes, return the result immediately
        if p.next == None:
            return head

        #Loop through next nodes
        while p.next != None:
            #If only one node left, return head
            if p.next.next == None:
                return head

            q = p.next
            t = p.next.next

            q.next = t.next
            t.next = q
            p.next = t

            p = q

        return head

File: code/ex24.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* p = dummy;
        while (p->next != NULL && p->next->next != NULL) {
            ListNode* n1 = p->next;
            ListNode* n2 = n1->next;
            n1->next = n2->next;
            p->next = n2;
            n2->next = n1;
            p = n1;
        }
        return dummy->next;
    }
};